pkg/runtime/serializer/codec_factory.go

```diff
@@ -17,36 +17,36 @@ limitations under the License.
 package serializer
 
 import (
+	"io/ioutil"
+
 	"k8s.io/kubernetes/pkg/api/unversioned"
 	"k8s.io/kubernetes/pkg/runtime"
 	"k8s.io/kubernetes/pkg/runtime/serializer/json"
 	"k8s.io/kubernetes/pkg/runtime/serializer/recognizer"
+	"k8s.io/kubernetes/pkg/runtime/serializer/streaming"
 	"k8s.io/kubernetes/pkg/runtime/serializer/versioning"
 )
 
+// serializerExtensions are for serializers that are conditionally compiled in
+var serializerExtensions = []func(*runtime.Scheme) (serializerType, bool){}
+
 type serializerType struct {
 	AcceptContentTypes []string
 	ContentType        string
 	FileExtensions     []string
-	Serializer         runtime.Serializer
-	PrettySerializer   runtime.Serializer
-}
 
-// NewCodecFactory provides methods for retrieving serializers for the supported wire formats
-// and conversion wrappers to define preferred internal and external versions. In the future,
-// as the internal version is used less, callers may instead use a defaulting serializer and
-// only convert objects which are shared internally (Status, common API machinery).
-// TODO: allow other codecs to be compiled in?
-// TODO: accept a scheme interface
-func NewCodecFactory(scheme *runtime.Scheme) CodecFactory {
-	return newCodecFactory(scheme, json.DefaultMetaFactory)
+	Serializer       runtime.Serializer
+	PrettySerializer runtime.Serializer
+	// If set, will be used for streaming codecs. This serializer is assumed to not need explicit
+	// typing via a wrapper.
+	ImplicitSerializer runtime.Serializer
```

Comment:
Serializer wraps the raw proto object (runtime.Object) with a
runtime.Unknown.  ImplicitSerializer wouldn't.

As an alternative, we could add an options struct to the
EncodeToStream/Decode methods that had a flag for that.  However, that's a
more drastic change, and only a few places need ImplicitSerializer.
UnwrappedSerializer or RawSerializer are other names.

On Mon, Mar 7, 2016 at 7:33 AM, Wojciech Tyczynski <notifications@github.com
> wrote:

> In pkg/runtime/serializer/codec_factory.go
> <https://github.com/kubernetes/kubernetes/pull/20377#discussion_r55202846>
> :
>
> >
> > -// NewCodecFactory provides methods for retrieving serializers for the supported wire formats
> > -// and conversion wrappers to define preferred internal and external versions. In the future,
> > -// as the internal version is used less, callers may instead use a defaulting serializer and
> > -// only convert objects which are shared internally (Status, common API machinery).
> > -// TODO: allow other codecs to be compiled in?
> > -// TODO: accept a scheme interface
> > -func NewCodecFactory(scheme *runtime.Scheme) CodecFactory {
> > -	return newCodecFactory(scheme, json.DefaultMetaFactory)
> > +	Serializer       runtime.Serializer
> > +	PrettySerializer runtime.Serializer
> > +	// If set, will be used for streaming codecs. This serializer is assumed to not need explicit
> > +	// typing via a wrapper.
> > +	ImplicitSerializer runtime.Serializer
>
> I mean, why can't we use "Serializer" here.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20377/files#r55202846>.
>